pkg/controller/servicebindingrequest/servicebinder.go

```diff
@@ -236,6 +241,64 @@ func (b *serviceBinder) onError(
 	return requeueOnNotFound(err, requeueAfter)
 }
 
+// isApplicationSelectorEmpty returns true if applicationSelector is not declared in
+// the Service Binding Request.
+func isApplicationSelectorEmpty(
+	application v1alpha1.ApplicationSelector,
+) bool {
+	var emptyApplication v1alpha1.ApplicationSelector
+	if application == emptyApplication ||
+		application.ResourceRef == "" &&
+			application.LabelSelector.MatchLabels == nil {
+		return true
+	}
+	return false
+}
+
+func addFinalizer(sbr *v1alpha1.ServiceBindingRequest) {
+	sbr.SetFinalizers(append(removeStringSlice(sbr.GetFinalizers(), finalizer), finalizer))
+}
+
+func removeFinalizer(sbr *v1alpha1.ServiceBindingRequest) {
+	sbr.SetFinalizers(removeStringSlice(sbr.GetFinalizers(), finalizer))
+}
+
+// handleApplicationError handles scenarios when:
+// 1. applicationSelector not declared in the Service Binding Request
+// 2. application not found
+func (b *serviceBinder) handleApplicationError(reason string, applicationError error, sbrStatus *v1alpha1.ServiceBindingRequestStatus) (reconcile.Result, error) {
+	conditionsv1.SetStatusCondition(&sbrStatus.Conditions, conditionsv1.Condition{
+		Type:    InjectionReady,
+		Status:  corev1.ConditionFalse,
+		Reason:  reason,
+		Message: applicationError.Error(),
+	})
+
+	// updating status of request instance
+	sbr, err := b.updateStatusServiceBindingRequest(b.sbr, sbrStatus)
+	if err != nil {
+		return requeueError(err)
+	}
+
+	// appending finalizer, should be later removed upon resource deletion
+	addFinalizer(b.sbr)
+	if _, err = b.updateServiceBindingRequest(sbr); err != nil {
+		return requeueError(err)
+	}
+
+	b.logger.Info(applicationError.Error())
+
+	if errors.Is(applicationError, errApplicationNotFound) {
+		removeFinalizer(b.sbr)
+		if _, err = b.updateServiceBindingRequest(sbr); err != nil {
+			return requeueError(err)
+		}
+		return requeueOnNotFound(errApplicationNotFound, requeueAfter)
```

Comment:
The original line indicated to always requeue on the scenario an application can't be found, so the following might make the intention clearer:

```suggestion
		return requeue(applicationError, requeueAfter)
```

Reply:
Why not use requeueOnNotFound(...) ?

```suggestion
		return requeueOnNotFound(applicationError, requeueAfter)
```

Reply:
Your suggestion is equivalent to my suggestion within the condition wrapping this return statement, since within the `if errors.Is(applicationError, errApplicationNotFound) {` block `applicationError` will always itself be or be wrapped by an `errApplicationNotFound` error.

Reply:
addressed 1f5f505